connect/api/src/main/java/org/apache/kafka/connect/storage/Converter.java

```diff
@@ -44,11 +45,34 @@
      */
     byte[] fromConnectData(String topic, Schema schema, Object value);
 
+    /**
+     * Convert a Kafka Connect data object to a native object for serialization.
+     * @param topic the topic associated with the data
+     * @param headers the headers associated with the data
+     * @param schema the schema for the value
+     * @param value the value to convert
+     * @return the serialized value
+     */
+    default byte[] fromConnectData(String topic, Headers headers, Schema schema, Object value) {
```

Comment:
They aren't, but they could be needed as metadata for serializing the payload into a byte array.

Reply:
http headers are an example of how headers could affect this implementation (whether I think it is a good idea or not :)). http headers could indicate `Content-Type` and `Content-Encoding` headers, which would affect how the message body is decoded. extremely simple version of this would be if you decoded as ASCII, UTF-8, or something else.

there are compatibility concerns, and I think discussion as to the best interfaces to dealing with the successive translations of data in a multilayered messages (similar to http, and kafka w/ headers) is worthwhile. for the most part, kafka has treated each field independently but with more layers in the message, this is likely unrealistic.

i think careful evolution of these interfaces would be valuable. I think it probably *also* makes sense to think of this holistically -- core, connect, streams -- though I recognize the challenge and barrier to doing that. In particular, passing an entire `Record` object seems like a more uniform approach and one which requires less future evolution (e.g. how many `fromConnectData()` methods will we need?). However, I also recognize the challenge in unifying this type (across consumers, producers, connect source, connect sink, streams input/output).

That said, I think it is worth evaluating before allowing the set of APIs to sprawl as it might start to do here.

Reply:
Holistic approach sounds like a good option, however it's a bigger and riskier change. My motivation was existing [Serializer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/serialization/Serializer.java#L61-L63) and [Deserializer](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/serialization/Deserializer.java) interfaces, so I'm going to target my KIP to bring Converter to the same level.